### ISSUE TITLE
Attach Database SQLite

### ISSUE THREAD

**Issue by elmpp** [mikro-orm/mikro-orm#1180]
Hi there,

I would like to contribute support for [SQLite attach database](https://sqlite.org/lang_attach.html). This feature operates very similar to multi-database of other rdbms but using filepaths in place of credentials. 

I have opened a [functioning PR](https://github.com/typeorm/typeorm/pull/7137) with typeorm and would likely approach this in a similar fashion. 

Would i be ok to attempt a PR?

**Comment by B4nan** [comment]
Sure, but looking at the PR it's 10k LoC, so if you want the PR to be merged, please try to do something smaller so it is reviewable :] I see its mostly about some updates in dependencies - that should in general never be part of a PR (if something requires dependency update, the update should be done separately).

Ideally start with summing up what changes would you like to do (from the public API standpoint). Tried to scan the PR quickly, but the number of unrelated changes (mainly in the docs) is too huge to find it :]

**Comment by elmpp** [comment]
Yes, there were many subsequent changes that aren't really relevant to this discussion. This [comparison view](https://github.com/typeorm/typeorm/compare/master...e740d2d91b4f640c9066f3818142873e5431c3ac) is best (just ignore docs/faq)

**Comment by B4nan** [comment]
> One thing i probably would say though is that this is probably best left til after foreign key migrations are supported as foreign key constraints are also prefixable

Looks like the blocker on knex's side for FKs will be fixed soon: https://github.com/knex/knex/pull/4162

Will work on the missing FKs support right after this is merged and released.

**Comment by Kristonitas** [comment]
> Will work on the missing FKs support right after this is merged and released.

[Its done!](https://github.com/knex/knex/pull/4162#issuecomment-751377981)

> Released in 0.21.15

**Comment by B4nan** [comment]
Well, done but not really. The issue I raised is still valid/opened and the use case I needed is still not possible to do with knex. But there is a PR that should help with that already.

**Comment by B4nan** [comment]
FYI the fix in knex was still not fixing the underlying issue, but I managed to get around that (as always, by monkey patching knex :trollface:). Things are in v5, as upgrading knex required to raise minimal TS version, but the FKs in sqlite are finally back, at least for the initial migrations (diffing is somewhat working too, but there is a lot of temp table magic and unfortunatelly with the way it works in knex I suspect there will be issues). 

So this is unblocked, master is v5 already, feel free to give it a try.

**Comment by elmpp** [comment]
Thanks @B4nan . I seem to recall you mentioning somewhere that you had a medium term goal of replacing Knex? I'm thinking it's best i wait til after that. It's also for the more selfish reason of my planning a Mikro migration in around the same timeframe

**Comment by B4nan** [comment]
Well, yes and no. I say this every time I see another blocker caused by knex :D Latest blocker is actually the sqlite diffing, as to generate down migrations it is unusable (knex always uses current schema when generating the temp table alter queries, which is obviously wrong in case of down migrations). 

I might do it in two steps, first get rid of it as schema generator dependency, second would be to reimplement pooling/tx handling and replace it from the runtime. 1. might happen before v5 is stable, but 2. probably not (as there are no real painpoints caused by that).

**Comment by elmpp** [comment]
It's never nice to get penned in by dependencies I'm sure you'd agree.
Library authors declaring classes as `final` in my PHP years used to wind
me up no end haha

As this would span schema, queries and migrations i'm gonna hold off then
as i say. Leave this open as i'd like to contribute this. It's proving very
useful with my use of sqlite.
